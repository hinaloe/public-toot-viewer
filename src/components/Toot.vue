<template>
    <div class="card mb-3">
        <div class="card-block d-flex">
            <div style="width: 90px;margin-right: 20px">
                <a :href="toot.account.url">
                    <img :src="toot.account.avatar" alt="avatar" style="border-radius: 50%;width: 100%">
                </a>
            </div>
            <div style="width: 100%">
                <div class="d-flex toot-header">
                    <a :href="toot.account.url" :title="note">
                        <div>{{toot.account.display_name}} @{{toot.account.acct}}</div>
                    </a>
                    <a :href="toot.url" class="ml-auto text-right date">{{date}}</a>
                </div>
                <p v-if="toot.spoiler_text">{{toot.spoiler_text}}
                    <button @click.prevent="more=true" class="btn btn-outline-primary btn-sm" v-if="!more">
                        Read more...
                    </button>
                    <button @click.prevent="more=false" class="btn btn-outline-secondary btn-sm" v-else>
                        Close
                    </button>
                </p>
                <div v-html="toot.content" v-if="!toot.spoiler_text||more"></div>
                <div v-if="toot.media_attachments.length">
                    <div v-if="toot.sensitive && !allowSensitive" class="text-center sensitive" @click="allowSensitive=true">
                        Sensitive content (click to show)
                    </div>
                    <div v-else>
                        <a :href="media.url" v-for="media in toot.media_attachments">
                            <img :src="media.preview_url" alt="" class="img-fluid img-thumbnail">
                        </a>
                    </div>
                </div>
            </div>
        </div>
    </div>
</template>
<style>
    .card .invisible {
        width: 0;
        height: 0;
        visibility: hidden;
        display: inline-flex;
    }
</style>
<style scoped="scoped">
    .card {
        word-break: break-all;
    }

    .sensitive {
        width: 100%;
        background: #303030;
        color: aliceblue;
        padding: 60px 30px;
        word-break: keep-all;
    }

    @media screen and (max-width: 768px) {
        .toot-header {
            flex-direction: column;
        }
    }
    .toot-header .date {
        font-size: .8em;
        color: #626262;
    }
</style>
<script>
  import striptags from 'striptags'
  export default {
    props: {
      toot: {
        type: Object, required: true
      }
    },
    data () {
      return {
        allowSensitive: false,
        more: false
      }
    },
    computed: {
      date () {
        const date = new Date(this.toot.created_at)
        return `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()} ${date.getHours()}:${`00${date.getMinutes()}`.slice(-2)}:${`0${date.getSeconds()}`.slice(-2)}`
      },
      note () {
        return striptags(this.toot.account.note.replace(/<br(?: \/)?>/g, '\n'))
      }
    }
  }
</script>
