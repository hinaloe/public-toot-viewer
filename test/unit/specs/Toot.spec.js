import Vue from 'vue'
import Toot from '@/components/Toot'

describe('Toot', () => {
  it('Toot render', () => {
    const Constructor = Vue.extend(Toot)
    const toot = {
      'id': 6069779,
      'created_at': '2017-04-30T19:37:51.090Z',
      'in_reply_to_id': null,
      'in_reply_to_account_id': null,
      'sensitive': false,
      'spoiler_text': '',
      'visibility': 'public',
      'application': {'name': 'Web', 'website': null},
      'account': {
        'id': 42621,
        'username': 'mikamiyoh',
        'acct': 'mikamiyoh',
        'display_name': '三上洋',
        'locked': false,
        'created_at': '2017-04-15T15:01:27.211Z',
        'followers_count': 1622,
        'following_count': 1679,
        'statuses_count': 1307,
        'note': '<p>ITジャーナリスト<br />マストドンは島→ <a href="https://news.yahoo.co.jp/byline/mikamiyoh/20170422-00070187/" rel="nofollow noopener" target="_blank"><span class="invisible">https://</span><span class="ellipsis">news.yahoo.co.jp/byline/mikami</span><span class="invisible">yoh/20170422-00070187/</span></a></p>',
        'url': 'https://mstdn.jp/@mikamiyoh',
        'avatar': 'https://media.mstdn.jp/images/accounts/avatars/000/042/621/original/b2b726b42455ea71.JPG',
        'avatar_static': 'https://media.mstdn.jp/images/accounts/avatars/000/042/621/original/b2b726b42455ea71.JPG',
        'header': '/headers/original/missing.png',
        'header_static': '/headers/original/missing.png'
      },
      'media_attachments': [],
      'mentions': [],
      'tags': [],
      'uri': 'tag:mstdn.jp,2017-04-30:objectId=6069779:objectType=Status',
      'content': '<p>すまんー。通話できひん。これから鳥キャスするの。</p>',
      'url': 'https://mstdn.jp/@mikamiyoh/6069779',
      'reblogs_count': 0,
      'favourites_count': 0,
      'reblog': null
    }
    const vm = new Constructor({propsData: {toot}}).$mount()
    expect((vm instanceof Vue)).to.equal(true)
    expect(vm.$el.querySelector('.toot-header a div').textContent)
      .to.equal('三上洋 @mikamiyoh')
  })
})
