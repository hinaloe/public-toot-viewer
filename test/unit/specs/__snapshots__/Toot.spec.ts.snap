// Vitest Snapshot v1

exports[`Toot > should render Toot 1`] = `
"<article data-v-cf45f0fa=\\"\\" class=\\"card mb-3\\">
  <div data-v-cf45f0fa=\\"\\" class=\\"card-body d-flex\\">
    <div data-v-cf45f0fa=\\"\\" style=\\"width: 90px; margin-right: 20px;\\"><a data-v-cf45f0fa=\\"\\" href=\\"https://mstdn.jp/@mikamiyoh\\"><img data-v-cf45f0fa=\\"\\" src=\\"https://media.mstdn.jp/images/accounts/avatars/000/042/621/original/b2b726b42455ea71.JPG\\" alt=\\"avatar\\" style=\\"border-radius: 50%; width: 100%;\\" /></a></div>
    <div data-v-cf45f0fa=\\"\\" style=\\"width: 100%;\\">
      <div data-v-cf45f0fa=\\"\\" class=\\"d-flex toot-header\\"><a data-v-cf45f0fa=\\"\\" href=\\"https://mstdn.jp/@mikamiyoh\\" title=\\"ITジャーナリスト
マストドンは島→ https://news.yahoo.co.jp/byline/mikamiyoh/20170422-00070187/\\">
          <div data-v-cf45f0fa=\\"\\"><span data-v-cf45f0fa=\\"\\" class=\\"name\\">三上洋</span> <span data-v-cf45f0fa=\\"\\" class=\\"acct\\">@mikamiyoh</span></div>
        </a><a data-v-cf45f0fa=\\"\\" href=\\"https://mstdn.jp/@mikamiyoh/6069779\\" class=\\"ml-auto text-right date\\">2017/4/30 19:37:51</a></div>
      <!---->
      <div data-v-cf45f0fa=\\"\\" class=\\"content\\">
        <p>すまんー。通話できひん。これから鳥キャスするの。</p>
      </div>
      <!---->
    </div>
  </div>
</article>"
`;
